Probe Implementation

ReubenJ opened the pull request:
Implementation of the method from ["Just-in-time learning for bottom-up enumerative synthesis"](https://doi.org/10.1145/3428295).

<img width="766" alt="afbeelding" src="https://github.com/user-attachments/assets/1cae8697-e097-4f13-9eb0-a1e3261dfb92">


TODO:
1. [x] Init Probe top-down iterator
2. [x] init update function for the probabilities of the grammar
3. [x] write priority function that multiplies the priority function of the program and returns the most likely program
4. [x] update `derivation_heuristic` accordingly
5. [x] write tests

- [x] check that log probabilities align with the intended enumeration order
- [ ] re-use parent's probabilities to calculate the current probability again.

Possible future work:
- [ ] #8 ? 
- [x] Add BottomUp iterator for Probe

ReubenJ committed:
Add skeleton for probe

THinnerichs committed:
Add Probe draft

THinnerichs commented:
I added a prototypical implementation of Probe to Garden. I still have to write proper tests + docs.

@ReubenJ I am overwriting some core functions of Herb, like `priority_function`, `rulenode_log_probability,`... . Also, I am not happy with overwriting all top-down iterators and having to do the same for all possible iterator-type classes. 
Maybe there is a smoother way?

THinnerichs committed:
Fix Probe's priority value function, renamed to ProbabilisticTopDownIterator

THinnerichs committed:
Add MLFSIterator to PRobe

THinnerichs committed:
Merge main into Probe

THinnerichs committed:
Checkpoint for old CP4PS experiments

THinnerichs committed:
Use new intepret generator for probe

THinnerichs committed:
Update Probe to use new interpret genereator; use cost-based BUS; updated README

ReubenJ committed:
chore: drop format script, switch to pre-commit, format with runic

ReubenJ committed:
Merge branch 'main' into feat/probe

ReubenJ committed:
chore: remove extraneous run.jl script

ReubenJ committed:
chore: revert patch number to 1

ReubenJ committed:
chore: remove exports

The way that we suggest people to use the garden doesn't require
exports.

`using Garden.Probe: probe` works whether `Probe` and `probe` are
exported or not.

ReubenJ commented:
~All good on my end @THinnerichs~ 🤥

ReubenJ committed:
refactor: switch `using`s to `import`s

ReubenJ commented:
Ok, tests were broken because of the updates to imports/exports. Should be good now 🤞
